File: kotlin/src/main/kotlin/benchmarks/Pidigits.kt
package benchmarks

import Benchmark
import java.math.BigInteger

class Pidigits : Benchmark() {
    private lateinit var output: StringBuilder
    private var nn: Long = 0

    init {
        nn = configVal("amount")
    }

    override fun prepare() {
        output = StringBuilder()
    }

    override fun run(iterationId: Int) {
        var i = 0
        var k = 0
        var ns = BigInteger.ZERO
        var a = BigInteger.ZERO
        var t: BigInteger
        var u: BigInteger
        var k1 = BigInteger.ONE
        var n = BigInteger.ONE
        var d = BigInteger.ONE

        while (true) {
            k += 1
            t = n.shiftLeft(1)
            n = n.multiply(BigInteger.valueOf(k.toLong()))
            k1 = k1.add(BigInteger.valueOf(2))
            a = a.add(t).multiply(k1)
            d = d.multiply(k1)
            if (a >= n) {
                val temp = n.multiply(BigInteger.valueOf(3)).add(a)
                val divResult = temp.divideAndRemainder(d)
                t = divResult[0]
                u = divResult[1].add(n)

                if (d > u) {
                    ns = ns.multiply(BigInteger.TEN).add(t)
                    i += 1
                    if (i % 10 == 0) {
                        output.append(String.format("%010d\t:%d\n", ns.toLong(), i))
                        ns = BigInteger.ZERO
                    }
                    if (i >= nn) break
                    a = a.subtract(d.multiply(t)).multiply(BigInteger.TEN)
                    n = n.multiply(BigInteger.TEN)
                }
            }
        }

        if (ns != BigInteger.ZERO) {
            output.append(String.format("%010d\t:%d\n", ns.toLong(), i))
        }
    }

    override fun checksum(): UInt = Helper.checksum(output.toString())

    override fun name(): String = "Pidigits"
}

File: clojure/src/benchmarks/pidigits.clj
(ns benchmarks.pidigits
  (:require [benchmarks.core :refer [bench config-i64]]
            [benchmarks.helper :refer [checksum-str]])
  (:import [java.math BigInteger]
           [java.io ByteArrayOutputStream]))

(defn- compute-pidigits [nn baos]
  (loop [i 0
         k 0
         ns BigInteger/ZERO
         a BigInteger/ZERO
         k1 1
         n BigInteger/ONE
         d BigInteger/ONE]

    (if (>= i nn)

      (when (> (.compareTo ns BigInteger/ZERO) 0)
        (let [line (format (str "%0" (mod nn 10) "d\t:%d\n")
                          (.longValue ns) nn)]
          (.write baos (.getBytes line))))

      (let [k' (inc k)
            t (.shiftLeft n 1)
            n' (.multiply n (BigInteger/valueOf k'))
            k1' (+ k1 2)
            a' (.multiply (.add a t) (BigInteger/valueOf k1'))
            d' (.multiply d (BigInteger/valueOf k1'))]

        (if (>= (.compareTo a' n') 0)

          (let [temp (.add (.multiply n' (BigInteger/valueOf 3)) a')
                quotient (.divide temp d')
                remainder (.remainder temp d')
                u (.add remainder n')]

            (if (> (.compareTo d' u) 0)

              (let [ns' (.add (.multiply ns BigInteger/TEN) quotient)
                    i' (inc i)
                    a'' (.multiply (.subtract a' (.multiply d' quotient))
                                  BigInteger/TEN)
                    n'' (.multiply n' BigInteger/TEN)]

                (if (zero? (mod i' 10))

                  (let [line (format "%010d\t:%d\n" (.longValue ns') i')]
                    (.write baos (.getBytes line))
                    (recur i' k' BigInteger/ZERO a'' k1' n'' d'))  

                  (recur i' k' ns' a'' k1' n'' d')))  

              (recur i k' ns a' k1' n' d')))  

          (recur i k' ns a' k1' n' d'))))))  

(bench "Pidigits"
  (let [nn (atom 0)
        result (atom nil)])

  (init
    (reset! nn (config-i64 "Pidigits" "amount"))
    (reset! result (ByteArrayOutputStream.)))

  (run [_]
    (let [nn-val @nn
          baos @result]
      (compute-pidigits nn-val baos)))

  (checksum
    (checksum-str (.toString @result))))